### DISCUSSION TITLE
Slight differences in scan results compared to other apps

### DISCUSSION THREAD

**Discussion by garuffalo** [complexlogic/rsgain#101] [Q&A]
I did some comparisons of replaygain scan results from different applications, and found that they were sometimes exactly the same as rsgain, but sometime had slight differences -- differences too slight to make any real world difference, but varying enough to stir my curiosity.

I scanned the same album of m4a files using rsgain and DeaDBeeF (both Linux), and foobar2000 and MusicBee (both Windows).

• DeaDBeeF had the exact same results as rsgain (rsgain easy with no clipping protection set in a profile).

• MusicBee and foobar2000 produced identical results, but some values, e.g., album peak, were off slightly from rsgain. Per track adjustments were virtually identical, with only slight difference that I suppose might be from some rounding convention (a guess).

From what I gather, all these applications use the same type of scanner, based on the EBU R 128 standard, although using -18 LUFS.

What accounts for the very small differences in scan results?
![deadbeef_replaygain_PBCB](https://github.com/complexlogic/rsgain/assets/28717740/2f99d7c7-48f1-40ba-a5a1-9ef4ecd111ee)
![replaygain_foobar2000_PBCB](https://github.com/complexlogic/rsgain/assets/28717740/f987998a-31c7-4a50-ae01-f53cbfb11d62)
![rsgain_PBCB_wprofile](https://github.com/complexlogic/rsgain/assets/28717740/ecc70cae-6cd7-4ff6-807a-0b088b1c346d)

**Comment by complexlogic** [comment]
Do those m4a files contain AAC?

**Reply by garuffalo** [reply]
Yes, they contain AAC. I should have mentioned that. I remember now that ALAC files have the same extension, but these are most certainly lossy, AAC.

**Reply by complexlogic** [reply]
Lossy codecs such as AAC generally use a floating point sample format. For lossy-encoded audio signals, the peak values often differ slightly depending on the decoder that is used. This is normally not significant enough to affect the gain calculations.

Based on your results, I presume that deadbeef and rsgain calculate the same peak values because they use the same decoder for AAC (FFmpeg "native" AAC decoder). foobar2000 and Musicbee are most likely using a different decoder such as FDK AAC. Those are closed source programs, so I'm not sure exactly what they are doing.

Try re-running your experiment with files that are encoded in a lossless codec, such as FLAC. You should expect to see identical peak values between the various programs, because a lossless file always decodes to a bit-identical copy of the original audio, regardless of the decoder used.

The difference in gain values between foobar2000/rsgain and deadbeef is likely just rounding.

**Reply by garuffalo** [reply]
Thank you for your thorough and clear explanation; also for your quick response to my question. Everything I have experienced about rsgain has been positive -- from the performance of the app, to the excellent documentation on Github, to the response of the developer. Great work! Cheers!

**Comment by sergeevabc** [comment]
Foobar2000 2.1 reports LUFS -8.90 and TP +2.20 (1.287598), but 

``` console
$ rsgain.exe custom -s s -t test.mp3
[OK] Scanning 'test.mp3'
[OK] Container: MP2/3 (MPEG audio layer 2/3) [mp3]
[OK] Stream #0: MP3 (MPEG audio layer 3), 44 100 Hz, 2 ch

...
Track: test.mp3
  Loudness:    -8.90 LUFS
  Peak:     1.178432 (1.43 dB)
  Gain:        -9.10 dB
```

Why 1.178432 (1.43 dB)?
